src/unity/toolkits/activity_classification/activity_classifier.cpp

```diff
@@ -173,41 +181,76 @@ void activity_classifier::init_options(
 std::tuple<float, float> activity_classifier::compute_validation_metrics(
     size_t prediction_window, size_t num_classes, size_t batch_size) {
 
+
   float cumulative_val_loss = 0.f;
   size_t val_size = 0;
   float cumulative_val_accuracy = 0.f;
   validation_data_iterator_->reset();
 
+  // To support double buffering, use a queue of pending inference results.
+  std::queue<result> pending_batches;
+
+  auto pop_until_size = [&](size_t remaining) {
+
+    while (pending_batches.size() > remaining) {
+
+      // Pop one batch from the queue.
+      result batch = pending_batches.front();
+      pending_batches.pop();
+
+      cumulative_val_accuracy += cumulative_chunk_accuracy(prediction_window, num_classes, batch.output_info, batch.data_info);
+
+      float val_loss = std::accumulate(batch.loss_info.data(), batch.loss_info.data() + batch.loss_info.size(),
+                                     0.f, std::plus<float>());
+      cumulative_val_loss += val_loss;
+
+    }
+  };
+
   while (validation_data_iterator_->has_next_batch()) {
+
     data_iterator::batch val =
         validation_data_iterator_->next_batch(batch_size);
+    
+    // Wait until we have just one asynchronous batch outstanding. The work
+    // below should be concurrent with the neural net inference for that batch.
+    pop_until_size(1);
+
+    result result_batch;
+
+    result_batch.data_info = val;
+
+
+    // Submit the batch to the neural net model.
     std::map<std::string, shared_float_array> results =
         training_model_->predict({
             {"input", val.features},
             {"labels", val.labels},
             {"weights", val.weights},
         });
 
-    const shared_float_array &output = results.at("output");
-    const shared_float_array &loss = results.at("loss");
-
-    float val_loss = std::accumulate(loss.data(), loss.data() + loss.size(),
-                                     0.f, std::plus<float>());
+    result_batch.output_info = results.at("output");
+    result_batch.loss_info = results.at("loss");
 
-    cumulative_val_loss += val_loss;
-
-    cumulative_val_accuracy +=
-        cumulative_chunk_accuracy(prediction_window, num_classes, output, val);
 
     val_size += val.batch_info.size();
+  
+    // Add the pending result to our queue and move on to the next input batch.
+    pending_batches.push(std::move(result_batch));
+    val = validation_data_iterator_->next_batch(batch_size);   
```

Comment:
I don't think we want this line, since you already fetch the next batch at the top of the loop body. As is, we're throwing away the result of this call!